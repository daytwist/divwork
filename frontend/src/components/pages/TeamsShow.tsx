import { FC, SyntheticEvent, useState } from "react";
import { Link } from "react-router-dom";
import {
  Avatar,
  Divider,
  Typography,
  Stack,
  Button,
  Box,
  Tabs,
  Tab,
} from "@mui/material";
import Grid2 from "@mui/material/Unstable_Grid2/Grid2";
import { PriorityBarChart } from "../models/team/PriorityBarChart";
import { TasksNewButton } from "../models/task/TasksNewButton";
import { LoadingColorRing } from "../ui/LoadingColorRing";
import { useFetchTeam } from "../../hooks/useFetchTeam";
import { TabPanel } from "../ui/TabPanel";
import { DeadlineBarChart } from "../models/team/DeadlineBarChart";

const TeamsShow: FC = () => {
  const { team, users } = useFetchTeam();
  const [tabValue, setTabValue] = useState(0);

  const tabProps = (index: number) => {
    return {
      id: `tab-${index}`,
      "aria-controls": `tabpanel-${index}`,
    };
  };

  const totals: number[] = [];
  // eslint-disable-next-line array-callback-return
  users?.map((user) => {
    const total = user.unfinished_tasks_priority_count.reduce(
      (sum: number, element: number) => {
        return sum + element;
      },
      0
    );
    totals.push(total);
  });
  const maxCount: number = Math.max(...totals);

  const handleSwitchTab = (event: SyntheticEvent, newValue: number) => {
    setTabValue(newValue);
  };

  return (
    <div>
      {team ? (
        <Grid2 container direction="column" rowSpacing={1}>
          <Grid2 xs={12}>
            <Typography
              gutterBottom
              variant="h4"
              component="div"
              data-testid="teams-show-h4"
            >
              {team.name}のタスク
            </Typography>
          </Grid2>
          <Grid2 xs={12}>
            <Stack
              direction="row"
              justifyContent="space-between"
              alignItems="center"
              mb={1}
            >
              <TasksNewButton />
              <Box sx={{ borderBottom: 1, borderColor: "divider" }}>
                <Tabs
                  value={tabValue}
                  onChange={handleSwitchTab}
                  textColor="inherit"
                >
                  <Tab label="優先度" {...tabProps(0)} />
                  <Tab label="納期" {...tabProps(1)} />
                </Tabs>
              </Box>
            </Stack>
          </Grid2>
          {users?.map((user) => (
            <Grid2 xs={12} key={user.id}>
              <Stack
                direction="row"
                justifyContent="space-between"
                alignItems="center"
                mb={1}
              >
                <Link
                  to={`/users/${user.id}`}
                  style={{ textDecoration: "none" }}
                >
                  <Stack
                    component={Button}
                    direction="column"
                    justifyContent="center"
                    alignItems="center"
                    sx={{
                      width: { xs: 80, sm: 100 },
                      height: { xs: 80, sm: 120 },
                      mr: { xs: 0, md: 2 },
                    }}
                  >
                    <Avatar
                      src={user.avatar}
                      alt="avatar"
                      sx={{
                        width: { sm: 60 },
                        height: { sm: 60 },
                      }}
                    />
                    <Typography
                      variant="button"
                      sx={{
                        color: "black",
                        mt: 1,
                      }}
                    >
                      {user.name}
                    </Typography>
                  </Stack>
                </Link>
                <Box
                  sx={{
                    width: { xs: 260, sm: 420, md: 530, lg: 750, xl: 900 },
                  }}
                >
                  <TabPanel value={tabValue} index={0}>
                    <PriorityBarChart user={user} maxCount={maxCount} />
                  </TabPanel>
                  <TabPanel value={tabValue} index={1}>
                    <DeadlineBarChart user={user} maxCount={maxCount} />
                  </TabPanel>
                </Box>
              </Stack>
              <Divider />
            </Grid2>
          ))}
        </Grid2>
      ) : (
        <LoadingColorRing />
      )}
    </div>
  );
};

export default TeamsShow;
