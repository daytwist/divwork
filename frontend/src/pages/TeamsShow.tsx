import { useState, useEffect, FC } from "react";
import { useParams, Link } from "react-router-dom";
import Cookies from "js-cookie";
import { Avatar, Button, Divider, Grid, Typography } from "@mui/material";
import { AxiosRequestConfig, AxiosResponse } from "axios";
import { axiosInstance } from "../utils/axios";
import { Team, TeamsShowResponse, User } from "../types";
import { TasksBar } from "../components/TasksBar";

const TeamsShow: FC = () => {
  const [team, setTeam] = useState<Team>();
  const [users, setUsers] = useState<User[]>();

  const params = useParams<{ id: string }>();

  const options: AxiosRequestConfig = {
    url: `/teams/${params.id}`,
    method: "GET",
    headers: {
      "access-token": Cookies.get("_access_token") || "",
      client: Cookies.get("_client") || "",
      uid: Cookies.get("_uid") || "",
    },
  };

  useEffect(() => {
    axiosInstance(options)
      .then((res: AxiosResponse<TeamsShowResponse>) => {
        console.log(res);
        setTeam(res.data.team);
        setUsers(res.data.users);
      })
      .catch((err) => {
        console.log(err);
      });
  }, []);

  const totals: number[] = [];
  // eslint-disable-next-line array-callback-return
  users?.map((user) => {
    const total = user.unfinished_tasks_count.reduce(
      (sum: number, element: number) => {
        return sum + element;
      },
      0
    );
    totals.push(total);
  });
  const maxCount: number = Math.max(...totals);

  return (
    <div>
      <Grid container direction="column" spacing={2}>
        <Grid item>
          <Typography
            gutterBottom
            variant="h4"
            component="div"
            data-testid="teams-show-h4"
          >
            {team?.name ? `${team.name}のタスク` : ""}
          </Typography>
        </Grid>
        {users?.map((user) => (
          <Grid item key={user.id}>
            <Grid
              container
              direction="row"
              justifyContent="space-between"
              alignItems="center"
            >
              <Grid item>
                <Grid
                  container
                  direction="column"
                  justifyContent="center"
                  alignItems="center"
                  sx={{
                    width: { xs: 80, sm: 100 },
                    height: { xs: 80, sm: 120 },
                    mr: { xs: 1, sm: 2 },
                  }}
                >
                  {user.avatar ? (
                    <Avatar
                      src={user.avatar}
                      alt="avatar"
                      sx={{
                        width: { sm: 60 },
                        height: { sm: 60 },
                      }}
                    />
                  ) : (
                    <Avatar
                      sx={{
                        width: { sm: 60 },
                        height: { sm: 60 },
                      }}
                    />
                  )}
                  <Button
                    variant="text"
                    size="large"
                    component={Link}
                    to={`/users/${user.id}`}
                    sx={{ p: { xs: 0.5 } }}
                  >
                    {user.name}
                  </Button>
                </Grid>
              </Grid>
              <Grid item>
                <TasksBar user={user} maxCount={maxCount} />
              </Grid>
            </Grid>
            <Divider />
          </Grid>
        ))}
      </Grid>
    </div>
  );
};

export default TeamsShow;
