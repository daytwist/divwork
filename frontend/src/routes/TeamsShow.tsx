import { useState, useEffect } from "react";
import { useParams } from "react-router-dom";
import { AxiosError, AxiosRequestConfig, AxiosResponse } from "axios";
import { axiosInstance } from "../utils/axios";
import { Team, TeamsShowResponse } from "../interfaces";

const TeamsShow: React.FC = () => {
  const [team, setTeam] = useState<Team>();
  const params = useParams<{ id: string }>();
  const options: AxiosRequestConfig = {
    url: `/teams/${params.id}`,
    method: "GET",
  };

  useEffect(() => {
    axiosInstance(options)
      .then((res: AxiosResponse<TeamsShowResponse>) => {
        const { data } = res;
        console.log(data);
        console.log(params);
        setTeam(data.team);
      })
      .catch((e: AxiosError<{ error: string }>) => {
        console.log(e.message);
      });
  }, []);

  return (
    <div>
      <h1>Teams Show</h1>
      <h2>{params.id}</h2>
      <h2>{team?.name}</h2>
    </div>
  );
};

export default TeamsShow;
