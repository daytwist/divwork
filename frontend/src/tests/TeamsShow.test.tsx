import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer"; // eslint-disable-line import/no-extraneous-dependencies
import { BrowserRouter } from "react-router-dom";
import TeamsShow from "../pages/TeamsShow";

describe("TeamsShow", () => {
  it("'Teams#Show'が表示されていること", () => {
    render(
      <BrowserRouter>
        <TeamsShow />
      </BrowserRouter>
    );
    const textElement = screen.getByText("Teams#Show");
    expect(textElement).toBeInTheDocument();
  });

  it("スナップショット", () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <TeamsShow />
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
