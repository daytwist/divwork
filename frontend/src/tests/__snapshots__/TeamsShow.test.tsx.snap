// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamsShow スナップショット 1`] = `
<div>
  <h1>
    Teams#Show
  </h1>
  <h2 />
  <div>
    <ul />
  </div>
</div>
`;
